References
==========

.. [Bowtie] Langmead et al., `Ultrafast and memory-efficient alignment of short DNA sequences to the human genome <http://genomebiology.com/2009/10/3/r25>`_, Genome Biology (2009). `Software homepage <http://bowtie-bio.sourceforge.net/index.shtml>`_.
.. [Cufflinks] (1) Trapnell et al., `Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation <http://www.nature.com/nbt/journal/v28/n5/full/nbt.1621.html>`_, Nature Biotechnology (2010). (2) Roberts et al., `Improving RNA-Seq expression estimates by correcting for fragment bias <http://genomebiology.com/2011/12/3/R22/abstract>`_, Genome Biology (2011). `Software homepage <http://cufflinks.cbcb.umd.edu>`_.
.. [FluxSimulator] Griebel et al., `Modelling and simulating generic RNA-Seq experiments with the flux simulator <http://nar.oxfordjournals.org/content/40/20/10073>`_, Nucleic Acids Research (2012). `Software homepage <http://sammeth.net/confluence/display/SIM/Home>`_.
.. [TopHat] Kim et al., `TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions <http://genomebiology.com/2013/14/4/R36/abstract>`, Genome Biology (2013). `Software homepage <http://tophat.cbcb.umd.edu/index.shtml>`. 
